connector/connect/common/src/main/protobuf/spark/connect/base.proto

```diff
@@ -435,6 +438,16 @@ message ExecutePlanResponse {
     // the execution is complete. If the server sends onComplete without sending a ResultComplete,
     // it means that there is more, and the client should use ReattachExecute RPC to continue.
   }
+
+  // This message is used to communicate progress about the query progress during the execution.
+  message ExecutionProgress {
+    int64 num_tasks = 1;
+    int64 num_completed_tasks = 2;
```

Comment:
Across all stages. It can always be extended later.

Reply:
I'm wondering how can this be accurate. With AQE we never know what is the number of partitions for the next stage, as re-optimization can happen.

Reply:
The goal of the progress metrics is not to be accurate into the future but only represent the snapshot of the current state. This means that the number of tasks can be updated when new stages are added or AQE kicks in.

The point is that the number of remaining tasks will converge over time and become stable.

Reply:
![Uploading progres.gif…]()

Reply:
![progres](https://github.com/apache/spark/assets/3421/709d2f3b-6af7-453b-b1e5-4b3e9e4c6ed1)

Reply:
This looks a bit different from what we have in classic Spark now. Classic Spark refreshes the progress bar after one stage is finished. So people may see something like below at the beginning
```
[Stage 1:=================================>                 (6619 + 12) / 10000]
```
then after stage 1 is completed, the progress bar is refreshed
```
[Stage 2:=====================>                             (3214 + 12) / 8000]
```

We don't have to follow the exact display format, but this "stage by stage" approach is more nature fit of the actual query execution in Spark.

Reply:
I think you're overindexing on accuracy. The progress bar is an ephemeral piece that simply gives the user feedback about the current operations happening. This is true for the current progress bar as well it shows a current stage and will delete it when it gets canceled.

To the user, it does not matter if the percentage is accurate or not as long as it converges toward 100% at some point in time. Simply seeing that any progress is happening and the concurrency of it is a benefit to the user.

Reply:
I agree we don't have to be very accurate, but is there a reason that we need to have a different progress bar style for Spark Connect? Is it because the classic Spark "stage by stage" progress bar is bad, or it's hard to implement?

Reply:
(Just my 2c: I think having any progress bar is much better than none. The standard Spark progress bar has some ups and some downs, definitely having new progress bars appear isn't the most intuitive either. I think it's probably net better than one progress bar that gets longer, but I would much prefer having some progress bar now that we can extend later, perhaps as we get a better sense of how to incorporate AQE and future stages into the UX.)

Reply:
After a second thought, it's better to hide Spark internals (stages) to end users, and eventually we should only have one progress bar for the query. So the current PR is a good starting point.

However, this server-client protocol needs to be stable and we don't want to change the client frequently to improve the progress reporting. Can we define a minimum set of information we need to send to the client side to display the progress bar? I feel it's better to calculate the percentage at the server side.

Reply:
So I refactored the code to avoid closing any doors. I did not change the way the progress bar is displayed. However, I extended the progress message to capture the stage-wise information so other clients can decide independently how to present the information to the end user.

Reply:
+1 @cloud-fan what do you think about that? Capture stage-level info in the proto, but keep the display simple for now?

Reply:
yea this is more flexible. The proto message contains all the information and clients can do whatever they want.